safenode/src/protocol/mod.rs

```diff
@@ -12,3 +12,100 @@ pub mod error;
+///
+/// This is the mapping from the XOR name used
+/// by for example self encryption, or the libp2p `PeerId`,
+/// to the key used in the Kademlia DHT.
+/// All our xorname calculations shall be replaced with the `KBucketKey` calculations,
+/// for getting proximity/distance to other items (whether nodes or data).
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+pub enum NetworkAddress {
+    /// The NetworkAddress is representing a PeerId.
+    PeerId(Vec<u8>),
+    /// The NetworkAddress is representing a ChunkAddress.
+    ChunkAddress(ChunkAddress),
+    /// The NetworkAddress is representing a DbcAddress.
+    DbcAddress(DbcAddress),
+    /// The NetworkAddress is representing a ChunkAddress.
+    RegisterAddress(RegisterAddress),
+}
+
+impl NetworkAddress {
+    /// Return a `NetworkAddress` representation of the `ChunkAddress`.
+    pub fn from_chunk_address(chunk_address: ChunkAddress) -> Self {
+        NetworkAddress::ChunkAddress(chunk_address)
+    }
+
+    /// Return a `NetworkAddress` representation of the `DbcAddress`.
+    pub fn from_dbc_address(dbc_address: DbcAddress) -> Self {
+        NetworkAddress::DbcAddress(dbc_address)
+    }
+
+    /// Return a `NetworkAddress` representation of the `RegisterAddress`.
+    pub fn from_register_address(register_address: RegisterAddress) -> Self {
+        NetworkAddress::RegisterAddress(register_address)
+    }
+
+    /// Return a `NetworkAddress` representation of the `PeerId` by encapsulating its bytes.
+    pub fn from_peer(peer_id: PeerId) -> Self {
+        NetworkAddress::PeerId(peer_id.to_bytes())
+    }
+
+    /// Return the encapsulated bytes of this `NetworkAddress`.
+    pub fn as_bytes(&self) -> Vec<u8> {
+        match self {
+            NetworkAddress::PeerId(bytes) => bytes.to_vec(),
+            NetworkAddress::ChunkAddress(chunk_address) => chunk_address.name().0.to_vec(),
+            NetworkAddress::DbcAddress(dbc_address) => dbc_address.name().0.to_vec(),
+            NetworkAddress::RegisterAddress(register_address) => register_address.id().0.to_vec(),
+        }
+    }
+
+    /// Return the `KBucketKey` representation of this `NetworkAddress`.
+    ///
+    /// The `KBucketKey` is used for calculating proximity/distance to other items (whether nodes or data).
+    /// Important to note is that it will always SHA256 hash any bytes it receives.
+    /// Therefore, the canonical use of distance/proximity calculations in the network
+    /// is via the `KBucketKey`, or the convenience methods of `NetworkAddress`.
+    pub fn as_kbucket_key(&self) -> Key<Vec<u8>> {
+        Key::new(self.as_bytes())
+    }
+
+    /// Compute the distance of the keys according to the XOR metric.
+    pub fn distance(&self, other: &NetworkAddress) -> Distance {
+        self.as_kbucket_key().distance(&other.as_kbucket_key())
+    }
+
+    // NB: Leaving this here as to demonstrate what we can do with this.
```

Comment:
I don't tthink we need this, specially because it's not using any of the `protocol` mechanisms but some kad fns.